text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1389 @@
+              str: string | string_dictionary 
+              i64: int64 
+              f64: float64 
+              bool: bool 
+              binary: binary | binary_dictionary
+              cbor: binary                                          # cbor encoded complex body value
+            attributes: *attributes
+            dropped_attributes_count: uint32 
+            flags: uint32
+```
+
+The type of the column `body` depends on the OTLP type and follows the same transformation rules used in the [attributes](#attribute-representation).
+
+#### Spans Payload
+
+The set of possible columns for a span payload is summarized in the following yaml description.
+
+```yaml
+resource_spans:
+  - resource: 
+      attributes: *attributes
+      dropped_attributes_count: uint32
+    schema_url: string | string_dictionary 
+    scope_spans: 
+      - scope: 
+          name: string | string_dictionary 
+          version: string | string_dictionary 
+          attributes: *attributes
+          dropped_attributes_count: uint32
+        schema_url: string | string_dictionary 
+        spans:                                                            # arrow type = list of struct
+          - start_time_unix_nano: timestamp                               # required 
+            end_time_unix_nano: timestamp                                 # required
+            trace_id: 16_bytes_binary | 16_bytes_binary_dictionary        # required, arrow fixed size binary array
+            span_id: 8_bytes_binary | 8_bytes_binary_dictionary           # required, arrow fixed size binary array
+            trace_state: string | string_dictionary 
+            parent_span_id: 8_bytes_binary | 8_bytes_binary_dictionary    # arrow fixed size binary array
+            name: string | string_dictionary                              # required
+            kind: uint8_dictionary                                        # OTLP enum with 6 variants 
+            attributes: *attributes
+            dropped_attributes_count: uint32 
+            events: 
+              - time_unix_nano: timestamp 
+                name: string | string_dictionary 
+                attributes: *attributes
+                dropped_attributes_count: uint32
+            dropped_events_count: uint32 
+            links: 
+              - trace_id: 16_bytes_binary | 16_bytes_binary_dictionary    # arrow fixed size binary array
+                span_id: 8_bytes_binary | 8_bytes_binary_dictionary       # arrow fixed size binary array
+                trace_state: string | string_dictionary 
+                attributes: *attributes
+                dropped_attributes_count: uint32 
+            dropped_links_count: uint32
+            status: 
+              code: uint8_dictionary                                      # OTLP enum with 4 variants 
+              status_message: string | string_dictionary
+```
+
+## Implementation Recommendations
+
```

Comment:
Interesting. I will read carefully this extension mechanism. I may come back to you in the future with some questions on this subject.

Reply:
Absolutely, currently there is some work being done on crafting some "Canonical Extension Types" or Well Known Types, which would be already included in most implementations. For example, a common Extension Type is an explicit UUID type which is just defined as a fixed_size_binary[16]. (You can see an example of this [here](https://github.com/apache/arrow/blob/go/v10.0.1/go/arrow/internal/testing/types/extension_types.go#L31)) So that *in the code* you can refer to a UUID array type, but the arrow data can pass through systems that have no knowledge of the UUID extension type without any issues.